### ISSUE TITLE
[java] BrokenNullCheck false positive with parameter/field confusion

### ISSUE THREAD

**Issue by maggu2810** [pmd/pmd#718]
**Rule:** [BrokenNullCheck](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#brokennullcheck)

**Description:**

The `BrokenNullCheck` of the basic rule set does not work as expected.
If I separate the condition of the `if` into separate statements the test succeeds.

The code sample below. The function `test1` and `test2` handle the same check. In `test1` boolean variables are created in `test2` there are no local variables created but we use the same statement in braces.

**Code Sample demonstrating the issue:**

```Java
public class PmdTest {

    public static boolean test1(final boolean hasFoo, final boolean hasBar, final Object foo, final Object bar) {
        final boolean hasFooObject = foo != null;
        final boolean hasBarObject = bar != null;
        if (hasFoo == hasFooObject && hasBar == hasBarObject) {
            return true;
        } else {
            return false;
        }
    }

    public static boolean test2(final boolean hasFoo, final boolean hasBar, final Object foo, final Object bar) {
        if (hasFoo == (foo != null) && hasBar == (bar != null)) {
            return true;
        } else {
            return false;
        }
    }
}
```

**Running PMD through:**

CLI

The error has been previously detected by a project using Maven build system. I created a small example and use the command line, so it could be tested easily.

Command line:

```sh
pmd-bin-5.8.1/bin/run.sh pmd -d PmdTest.java -f xml -l java -R rulesets/java/basic.xml
```

Output:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<pmd version="5.8.1" timestamp="2017-11-07T14:09:02.275">
<file name="/home/rathgeb/tmp/pmd/PmdTest.java">
<violation beginline="14" endline="18" begincolumn="9" endcolumn="9" rule="BrokenNullCheck" ruleset="Basic" class="PmdTest" method="test2" externalInfoUrl="https://pmd.github.io/pmd-5.8.1/pmd-java/rules/java/basic.html#BrokenNullCheck" priority="2">
Method call on object which may be null
</violation>
</file>
</pmd>
```

**Comment by jsotuyod** [comment]
@maggu2810 thanks for the very complete report. This is definitely a false positive. I don't believe this will make it into the upcoming 6.0.0 release, but definitely for a minor 6.x release.

**Comment by jsotuyod** [comment]
@SupunArunoda this is a false positive. As such, the issue relies solely on the rule's logic itself, that gets to flag a node it shouldn't.

So first things first, are you able to reproduce the issue by creating a new test case on `BrokenNullCheck.xml` such as this one?

https://github.com/pmd/pmd/blob/8aeddd09a41dff694d84a0ef0bbcf91df170d190/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/BrokenNullCheck.xml#L189-L202

If unable to reproduce we should check if we are missing anything / the original reporter can help us. If you can, then debugging the rule to look for the broken logic should be somewhat easier.

**Comment by SupunArunoda** [comment]
@jsotuyod ,
I added a test case related to this issue. But the test case is failed (Attached the screenshot). 
![test_failure](https://user-images.githubusercontent.com/12041588/35291203-9d6072c2-0092-11e8-9abf-38e10d9973c9.PNG)


But after adding another two brackets to comparison variables, test case has passed.
`public static boolean test2(final boolean hasFoo, final boolean hasBar, final Object foo, final Object bar)` 
     `{`  
         `if ((hasFoo == (foo != null)) && (hasBar == (bar != null))) {`
           `    return true;`
         `} else {`
          `     return false;`
        `}`
    `}`

**Comment by jsotuyod** [comment]
@SupunArunoda very interesting, but may simply point to another bug. I'd start without the extra parenthesis by debugging that test case and checking where it is we are falling through the wrong branch in the code.

**Comment by pzygielo** [comment]
I have encountered very similar case, so it might be the same cause.
```java
  private Boolean f;

  // parameter of the same name as field
  public void f(Boolean f) {
    if (f != null || this.f == null)
      ;   
  }
```
is reported as
```
[INFO] --- maven-pmd-plugin:3.13.0:check (default) @ ... ---
[INFO] PMD version: 6.28.0
[INFO] PMD Failure: ... Rule:BrokenNullCheck Priority:2 Method call on object which may be null.
```
<sup>(no method is being called at all - which is separate problem, just of message I think)</sup>.
I don't want to create duplicate, if it isn't separate case.

**Comment by iamankushpandit** [comment]
I am currently debugging this. I am able to reproduce the junit failure by adding the following
`<test-code>
        <description>ankush1</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
public class Foo {
     boolean bar(boolean hasFoo, boolean hasBar, Object foo, Object bar) {
        if (hasFoo == (foo != null) && hasBar == (bar != null)) {
            return true;
        } else {
            return false;
        }
    }
}
        ]]></code>
    </test-code>

`

From what I can tell so far is that in the checkForViolations method in the BrokenNullCheckRule.java, the following condition
`if (primaryExpressionsAreEqual(nullCompareExpression, conditionalPrimaryExpression)) {
                addViolation(data, node); // We have a match
            }`
adds the violation because both the nullCompareExpression and conditionalPrimaryExpression Nodes are "hasFoo".

Trying to understand it further.

The nullCompareExpression is computed by findNullCompareExpression method when equalityExpression is passed. Here equalityExpression is roughly "hasFoo == (foo != null)".

and conditionalPrimaryExpression is calculated by conditionalSubnode.getFirstDescendantOfType(ASTPrimaryExpression.class);
where conditionalSubnode is roughly "(hasFoo == (foo != null) && hasBar == (bar != null))"

I will continue to look in to this.

**Comment by iamankushpandit** [comment]
Looking further in to this, I feel that findNullCompareExpression should not return hasFoo but rather foo as in the statement hasFoo == (foo != null) the null compare expression is foo!=null and comparison of hasFoo to the value foo!=null happens later.

**Comment by iamankushpandit** [comment]
I changed the findNullCompareExpression method to not return the first primaryExpression but rather collect all in a result List and then return the last element of the result list which resolves this issue but breaks the Arrays are Ok test case which has the following code,

<![CDATA[
public class Foo {
    void bar() {
        if (index >= 0 && books[index] == null) { }
    }
}

Here the same logic of returning the last element of the result list fails as it returns 'index' and not books[index]. Looking in to finding a way to resolve it.
